### PR TITLE
Fix the logic to show "You are logged in for the first time" fragment

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
This PR fixes a bug when First time login fragment is shown if the same user logs out and logs in again on the same device and "You are logged in for the first time" dialog is still shown for him.

#### APK
[Download build #8817](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8817/artifact/build/artifact/wire-dev-PR835-8817.apk)

### CODE REVIEW

File: wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java [hunk 1]
```diff
@@ -510,7 +522,11 @@ private void setStateInternal(AppEntryState state, boolean forceUpdate) {
                 setRegistrationContext(RegistrationEventContext.PERSONAL_INVITE_PHONE);
                 break;
             case FIRST_LOGIN:
-                appEntryStateCallback.onShowFirstLaunchPage();
+                if (isFirstSignInOnDevice) {
```
Comment (JPonte) [inline wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java]: I think that if it's not the first login it shouldn't reach this state in the first place, so this check should be on the place where the `setState(AppEntryState .FIRST_LOGIN)` is.

Comment (mykola-mokhnach) [inline wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java]: I just didn't want to expose **isFirstSignInOnDevice** property. Also, state change has been already implemented like that if you check the previous implementation of _onShowFirstLaunchPage_

Comment (JPonte) [inline wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java]: What do you mean expose it? The `FIRST_LOGIN` state is being set on `EMAIL_SIGNED_IN` and `PHONE_SIGNED_IN_RESUMING` on that switch. It doesn't seem right to switch to the `FIRST_LOGIN` state if it's not in fact the first login.

Comment (mykola-mokhnach) [inline wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java]: not only. Check the EmailAddPhoneFragment->onNotNow method

Comment (JPonte) [inline wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java]: Right, makes sense then. I think it's a bit ugly to expose the setState outside that class but ok.

File: wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java [hunk 2]
```diff
@@ -605,6 +621,14 @@ public void onConfirmationCodeSendingFailed(KindOfAccess kindOfAccess,
 
     @Override
     public void setSignInPhone(final String countryCode, final String phone, final ErrorCallback errorCallback) {
+        try {
+            AccountsStorage storage = getAccountsStorage();
+            isFirstSignInOnDevice = storage == null || !Await.result(storage.findByPhone(phone),
+                Duration.create(1, TimeUnit.SECONDS)).isDefined();
+        } catch (Exception e) {
```
Comment (wire-android) [inline wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java]: What exception do you actually expect here? Catching unspecified exception is weird, it kind of says: "I have no idea what I'm doing, will catch all" :)
Besides that it seems that you are using exception to pass some logic, it's not recommended, would be better to check before if there will be exception.

File: wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java [hunk 3]
```diff
@@ -1045,8 +1069,23 @@ public void onRegistrationFailed(int i, String s, String s1) {
...
+        return global.accountsStorage();
+    }
+
     @Override
     public void signInWithEmail(String email, String password, final ErrorCallback errorCallback) {
+        try {
+            AccountsStorage storage = getAccountsStorage();
+            isFirstSignInOnDevice = storage == null ||
+                !Await.result(storage.findByEmail(email), Duration.create(1, TimeUnit.SECONDS)).isDefined();
+        } catch (Exception e) {
```
Comment (wire-android) [inline wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java]: again, catching anything

File: wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java (hunk 2 above)
Comment (mykola-mokhnach) [inline wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java]: This is the definition of Await.result:

```java
    public static <T> T result(Awaitable<T> var0, Duration var1) throws Exception {
}
```

can it be implemented differently?

Comment (wire-android) [inline wire-core/src/main/java/com/waz/zclient/core/api/scala/AppEntryStore.java]: oh, missed that;
waiting for db results on UI thread is even bigger issue :/ 
guess there is no easy way of implementing this with Await

**Comment by hakonbo** [comment]
Closing for now, please re-open when ready for merging

**Pull request closed without merging**